### PR TITLE
Implemented addBeanFromConstructor

### LINKED ISSUE #11317: Cleanup usages of `addBean(Object)` from constructors
**Jetty version(s)**
12

**Jetty Environment**
All

**Java version/vendor** `(use: java -version)`
All

**OS type/version**
All

**Description**
In the process of working Issue #11220 and the PR #11225 it was discovered that we have a problem with calling virtual methods (like `addBean(Object)`) from constructors where we can wind up with partially initialized classes that are then used / referenced / passed around in different ways.

One is the DEBUG logging of addBean with a parameter of `this` ...

https://github.com/jetty/jetty.project/blob/7fbd51a95233d492c2bbe85d77265cd2a9983799/jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/component/ContainerLifeCycle.java#L418-L419

Another is the call to the bean added listeners with `this` parameter during addBean.

https://github.com/jetty/jetty.project/blob/7fbd51a95233d492c2bbe85d77265cd2a9983799/jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/component/ContainerLifeCycle.java#L348-L350

We should strive to move to a doStart() > addBean() and later call to doStop() > removeBean() (like what is being done in https://github.com/jetty/jetty.project/pull/11225/commits/fc32100c3d03ab37cb756fc309767e7a8d7c5c86

### PR DESCRIPTION
**Pull request by gregw**
Added mechanism to safely add beans from a super constructor of ContainerLifeCycle

Fixes #11317

### CODE REVIEW

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/component/ContainerLifeCycle.java [hunk 1]
```diff
@@ -421,6 +415,53 @@ else if (isStarted())
...
+    /**
+     * Add a bean in a way that is safe to call from a super constructor of this {@code ContainerLifeCycle}:
+     * there are no {@link LifeCycle.Listener}s registered;
+     * the object itself is not a {@link LifeCycle.Listener};
+     * this {@link LifeCycle} is not started or starting;
+     * and the is no debugging call to {@code this.toString()}.
+     * @param o The bean to add
+     * @return true if the bean was added
+     */
+    protected boolean addBeanFromConstructor(Object o)
```
Comment (lorban) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/component/ContainerLifeCycle.java]: This method should be `final`, we absolutely do not want anyone to override it.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/component/ContainerLifeCycle.java [hunk 2]
```diff
@@ -421,6 +415,53 @@ else if (isStarted())
...
+     * Add a bean in a way that is safe to call from a super constructor of this {@code ContainerLifeCycle}:
+     * there are no {@link LifeCycle.Listener}s registered;
+     * the object itself is not a {@link LifeCycle.Listener};
+     * this {@link LifeCycle} is not started or starting;
+     * and the is no debugging call to {@code this.toString()}.
+     * @param o The bean to add
+     * @param managed true if the bean is to be managed.
+     * @return true if the bean was added
+     */
+    protected boolean addBeanFromConstructor(Object o, boolean managed)
```
Comment (lorban) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/component/ContainerLifeCycle.java]: ditto.

**Review by lorban — COMMENTED** [review]
I find some elegance in this proposal, the only things I'm reserved about are:

 - the `addBeanFromConstructor` method name is horrible, how about `installBean` instead?
 - maybe objects added with that new method should be impossible to remove with `removeBean`?
 - we should add a test to make sure `addBeanFromConstructor` doesn't make calls on the object-under-construction's reference, for instance by calling `toString` on it because of a logging call.
 - I'm reserving my final judgment for when all tests will be passing without any bad surprise.

**Comment by gregw** [comment]
> I find some elegance in this proposal, the only things I'm reserved about are:
> 
> * the `addBeanFromConstructor` method name is horrible, how about `installBean` instead?

OK

> * maybe objects added with that new method should be impossible to remove with `removeBean`?

We spend a lot of effort to make sure beans are kept up to date with update mechanisms, so I don't think this is needed.

> * we should add a test to make sure `addBeanFromConstructor` doesn't make calls on the object-under-construction's reference, for instance by calling `toString` on it because of a logging call.

working on it

> * I'm reserving my final judgment for when all tests will be passing without any bad surprise.

I've fixed some beans that were listeners, but I expect there will be a few more.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/component/ContainerLifeCycle.java [hunk 3]
```diff
@@ -421,6 +415,53 @@ else if (isStarted())
...
+            ? (managed ? Managed.MANAGED : Managed.UNMANAGED)
+            : (managed ? Managed.POJO : Managed.UNMANAGED));
+    }
+
+    private boolean installBean(Object o, Managed managed)
+    {
+        if (o == null || contains(o))
+            return false;
+        if (o instanceof LifeCycle.Listener || !_listeners.isEmpty())
+            throw new IllegalArgumentException("Cannot call Listeners from constructor");
```
Comment (sbordet) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/component/ContainerLifeCycle.java]: I'd remove these 2 lines, as there is no need to throw -- this method has a different semantic than `addBean()`.

If some code calls `addBean()` from a constructor, which adds a listener, and then calls `installBean()`, it should just work.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/component/ContainerLifeCycle.java [hunk 4]
```diff
@@ -421,6 +415,53 @@ else if (isStarted())
...
+
+        Bean newBean = new Bean(o);
+        newBean._managed = managed;
+        _beans.add(newBean);
+
+        if (LOG.isDebugEnabled())
+            LOG.debug("added from constructor {}", newBean);
```
Comment (sbordet) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/component/ContainerLifeCycle.java]:
```suggestion
            LOG.debug("{}@{} added {}", getClass().getSimpleName, hashCode(), newBean);
```

**Review by sbordet — CHANGES_REQUESTED** [review]

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/component/ContainerLifeCycle.java (hunk 3 above)
Comment (gregw) [reply jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/component/ContainerLifeCycle.java]: These lines were very useful in finding code calling installBean with an object that implements LifeCycle.Listener and expects to be called.
I think these line are key to enforcing the difference between these methods.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/component/ContainerLifeCycle.java [hunk 5]
```diff
@@ -311,13 +311,7 @@ public boolean isUnmanaged(Object bean)
     @Override
     public boolean addBean(Object o)
```
Comment (lorban) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/component/ContainerLifeCycle.java]: We should change the javadoc to explicitly state that `addBean` should not be called from constructors.

**Review by lorban — CHANGES_REQUESTED** [review]
While I think this change has merits, it does not address the original problem reported in https://github.com/jetty/jetty.project/issues/11220 as in that case, the cause was:

 - `Handler$Wrapper`'s constructor 
    - calls `Handler$Singleton.updateHandler`
      - calls `ContainerLifeCycle.updateBean`
        - calls `ContainerLifeCycle.addBean`

so `addBean` may be called indirectly from a general-purpose method, which bypasses this constructor-specific method.

**Comment by lorban** [comment]
Also, the following test fails when DEBUG logs are enabled, despite your change:

```java
    @Test
    public void testInstallBeanDoesNotEscapeConstructor()
    {
        ConcreteClazz c = new ConcreteClazz();
        assertThat(c.exception, is(nullValue()));
    }

    private static abstract class AbstractClazz extends ContainerLifeCycle
    {
        public AbstractClazz()
        {
            // Pass a reference to a half-constructed object to installBean.
            installBean(this);
        }
    }

    private static class ConcreteClazz extends AbstractClazz
    {
        private final List<String> aList = List.of("single element");

        private NullPointerException exception;

        @Override
        public String toString()
        {
            try
            {
                // Call a method on the aList field, if there was a constructor escape the ref will be null despite being final.
                String entry = aList.get(0);
                return "ConcreteClazz@" + hashCode() + "[" + entry + "]";
            }
            catch (NullPointerException npe)
            {
                exception = npe;
                throw npe;
            }
        }
    }
```

**Comment by gregw** [comment]
@lorban I think both of your examples are poor code:

+ The `Handler.Wrapper` should not be calling `updateHandler` from a constructor, as there is no handler to update.   I've factored out the sanity checks to a `checkHandler` method instead so now `installBean` can be used.

+ In the second example, I don't think it is valid to add a bean to itself.  Any constructor that passes this out in a method is vulnerable like that code.  But correct usage of `installBean` does not need this to be passed out.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/component/ContainerLifeCycle.java (hunk 3 above)
Comment (gregw) [reply jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/component/ContainerLifeCycle.java]: But actually it was the wrong listener.... changed to Container.Listener

**Comment by gregw** [comment]
@sbordet @lorban nudge

**Comment by gregw** [comment]
@sbordet @lorban nudge again

**Comment by gregw** [comment]
@sbordet @lorban nudge again again

File: jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/ContextHandler.java
Comment (lorban) [inline jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/ContextHandler.java:1]: The null checks on `_vhosts` in `getVirtualHosts()` and `toString()` introduced in https://github.com/jetty/jetty.project/pull/11318 should be reverted.

File: jetty-core/jetty-util/src/test/java/org/eclipse/jetty/util/component/LifeCycleListenerTest.java [hunk 6]
```diff
@@ -222,4 +223,58 @@ public void lifeCycleStopping(LifeCycle event)
             }
         }
     }
+
```
Comment (lorban) [inline jetty-core/jetty-util/src/test/java/org/eclipse/jetty/util/component/LifeCycleListenerTest.java:226]: The #1 motivation for this new `installBean` method was to make sure we don't de-ref objects under construction. We should find a way to test that.

**Review by lorban — CHANGES_REQUESTED** [review]
We need some test coverage to make sure we don't de-reference objects under construction, otherwise LGTM.

File: jetty-core/jetty-util/src/test/java/org/eclipse/jetty/util/component/LifeCycleListenerTest.java (hunk 6 above)
Comment (gregw) [reply jetty-core/jetty-util/src/test/java/org/eclipse/jetty/util/component/LifeCycleListenerTest.java:226]: Neither me nor chatGPT can think of a way that the method `addBean` can tell if it has been called from a super constructor or not.   Thus it is impossible for us to test if it is safe for `addBean` to call a listener passing this.     We have provided and alternative to `addBean` that is safe to call from constructors and we have javadoc'd that alternative.

I have visually scanned the list of callers of addBean and can see no constructors in there.  However, I cannot tell if any of those callers are themselves called from a constructor.

I doubt that we are perfect.... but we are better than we were!

**Comment by lorban** [comment]
I believe we should have added a test like the following as part of #11321. It is up to you to include it in this PR or not, as that is the best coverage I can come up with, and I agree it's not much.

```java
    @Test
    public void testHandlerWrapperDoesNotLogUnconstructedObject()
    {
        JettyLogger jettyLogger = (JettyLogger)LoggerFactory.getLogger(ContainerLifeCycle.class);
        JettyLevel originalLevel = jettyLogger.getLevel();
        PrintStream originalPrintStream = System.err;
        try
        {
            jettyLogger.setLevel(JettyLevel.ALL);
            ByteArrayOutputStream baos = new ByteArrayOutputStream();
            PrintStream printStream = new PrintStream(baos);
            System.setErr(printStream);

            new Handler.Wrapper(new EchoHandler())
            {
                private final List<String> strings = List.of("aString");

                @Override
                public String toString()
                {
                    // De-ref the final field here.
                    return "string[0]=" + strings.get(0);
                }
            };

            printStream.flush();
            String output = baos.toString(StandardCharsets.UTF_8);

            // Check that SLF4j did not print an error to stderr.
            assertThat(output, allOf(not(containsString("SLF4J: Failed toString()")), not(containsString("NullPointerException"))));
        }
        finally
        {
            jettyLogger.setLevel(originalLevel);
            System.setErr(originalPrintStream);
        }
    }
```

**Review by lorban — APPROVED** [review]

**Comment by gregw** [comment]
> I believe we should have added a test like the following as part of #11321. It is up to you to include it in this PR or not, as that is the best coverage I can come up with, and I agree it's not much.

Maybe as a full integration test with almost all our modules turned on, full debug and a few requests sent in, then we check the log for anything that should not be there.... But I think if it is just a few handlers, it does not test much.

**Comment by gregw** [comment]
@sbordet nudge!

**Review by sbordet — APPROVED** [review]

**Pull request merged**